insolver/feature_monitoring/homogeneity_report.py

```diff
@@ -0,0 +1,313 @@
+    """
+    This function draws histograms of given samples using joint grid.
+    It needs limits of interested area and number of bins.
+
+    Parameters:
+        x1 (np.array): sample from base period.
+        x2 (np.array): sample from current period.
+        name1 (str): name to describe base period.
+        name2 (str): name to describe current period.
+        offline (bool): whether to return plot for rendering in template or to return figure itself.
+
+    Returns:
+        plotly offline plot with histograms in case when 'offline' parameter is 'True'
+        else it returns plotly figure with histograms.
+    """
+
+    # draw discrete hists
+    fig1 = px.histogram(x1, histnorm='probability', barmode='overlay', color_discrete_sequence=['green'])
+    fig1.for_each_trace(lambda t: t.update(name=name1))
+    fig2 = px.histogram(x2, histnorm='probability', barmode='overlay', color_discrete_sequence=['red'])
+    fig2.for_each_trace(lambda t: t.update(name=name2))
+    fig = py.graph_objects.Figure(data=fig1.data + fig2.data)
+
+    # add details
+    fig.update_layout(
+        autosize=False,
+        width=830,
+        height=650,
+        legend=dict(x=0.8, y=0.95, traceorder='normal', font=dict(color='black', size=16)),
+    )
+
+    if offline:
+        return py.offline.plot(fig, include_plotlyjs=False, output_type='div')
+    else:
+        return fig
+
+
+class HomogeneityReport:
+    """
+    This class builds report for two dataframes containing same features.
+    Report consists of homogeneity checks between feature's condition in first and in second frame.
+    We run homogeneity tests for each feature and render report with distributions if necessary.
+
+    The class supports wide configuration of running tests. It takes config. dictionary with which
+    has feature names as keys. Each value for certain feature is sub-dictionary of properties.
+    List of supported properties:
+        feature_type (str): stat. type of features. ('continuous'/'discrete')
+        pval_thresh (float): threshold for pvalue for making conclusions in tests.
+        samp_size (int): size of sub-samples in bootstrap.
+        bootstrap_num (int): number of bootstrap restarts.
+        psi_bins (int): number of bins for psi calculation (only for 'continuous').
+        chart_bins (int): number of bins to draw hists (only for 'continuous').
+        chart_limits (iterable of two floats/ints): custom limits for continuous feature charts
+        ('only for continuous').
+
+    Only 'feature_type' ('continuous'/'discrete') must be specified for each feature,
+    other properties can be missing.
+
+    Default parameters for each property:
+    pval_thresh=0.05, samp_size=500, bootstrap_num=100,
+    psi_bins=20, chart_bins=15, chart_limits=[min(x1, x2), max(x1, x2)].
+
+    Parameters:
+        config_dict (dict): dict. with feature properties (see description).
+    """
+
+    def __init__(self, config_dict: dict = {}):
```

Comment:
Argument `config_dict` is mutable. Also, note that type annotations support only Python 3.8+.

Reply:
Decided to remove default val. for config_dict. Seems to be pointless when no features given.